docs/posts/ibis-duckdb-geospatial/index.qmd

```diff
@@ -0,0 +1,182 @@
+from ibis import _
+
+ibis.options.interactive = True
+
+con = ibis.duckdb.connect("nyc_data.db")
+con.list_tables()
+```
+
+In this database we have multiple tables with information about New York City. Following, Dr. Qiusheng Wu's class, we'll take a look at some spatial relations.
+
+We can start by taking a peak at the `nyc_subway_stations` table.
+
+```{python}
+subway_stations = con.table('nyc_subway_stations')
+subway_stations
+```
+
+Notice that the last column has a `geometry` type, and in this case it contains a points that represent the location of each subway station. Let's take the entry for `'Broad St'`.
+
+```{python}
+broad_station =  subway_stations.filter(subway_stations.NAME == 'Broad St')
+broad_station
+```
+
+### `geo_equals` (`ST_Equals`)
+
+In DuckDB `ST_Equals` returns `True` if two geometries are topologically equal, meaning that they have the same dimension and they have identical coordinate values, although the order of the vertices may be different.
+
+The following  is a bit redundant but we can check if our broad station point
+matches only one point on our data using `geo_equals`
+
+```{python}
+subway_stations.filter(subway_stations.geom.geo_equals(broad_station.geom))
+```
+
+We can also write this query without using `broad_station` as a variable, and with the help of the deferred expressions API, also known as Underscore (`_`) API (Learn more about it in the [chaining expressions docs](https://ibis-project.org/how-to/analytics/chain_expressions.html))
+
+```{python}
+subway_stations.filter(_.geom.geo_equals(_.filter(_.NAME == 'Broad St').geom))
+```
+
+### `intersect` (`ST_Intersect)
+
+Let's find in which neighborhood is "Broad Street" subway station and determine its neighborhood using the the geospatial `intersect` function. The `intersect` function returns `True` if two geometries have any points in common.
+
+The
+
+```{python}
+boroughs = con.table('nyc_neighborhoods')
+boroughs
+```
+
+
+### FIXME: NEED AN EXPLANATION ABOUT WHY
+it's still unclear to me why:
+a) this works broad_station.select(broad_station.geom).to_array())
+b) but this doesn't broad_station.geom.to_array()).  -> AttributeError: 'GeoSpatialColumn' object has no attribute 'to_array'
+
+Looks like a) is ibis.expr.types.relations.Table while b) ibis.expr.types.geospatial.GeoSpatialColumn
+find a way to explain why we do this
+
+```{python}
+boroughs.filter(boroughs.geom.intersects(broad_station.select(broad_station.geom).to_array()))
+```
+
+
+### `d_within` (ST_DWithin)
+We can also find what are the streets nearby (within 10 meters of) of the "Broad St" subway station using the `d_within` function. The `d_within` function returns True if the geometries are within a given distance.
+
+
+```{python}
+streets = con.table('nyc_streets')
+streets
+```
+
+Using the deferred API, we can check which streets are within `d=10` meters of distance.
+
+```{python}
+streets.filter(_.geom.d_within(broad_station.select(_.geom).to_array(), 10))
```

Comment:
How difficult would it be to generate a map using leaflet or some other library that hooks in to the IPython display mechanism?